benchmarks/distributed/rpc/ps/benchmark.py

```diff
@@ -0,0 +1,384 @@
+        super().__init__()
+        self.embedding = nn.EmbeddingBag(4, 4, sparse=True)
+        self.fc1 = nn.Linear(4, 4)
+
+    def forward(self, x):
+        x = self.embedding(x)
+        return F.softmax(self.fc1(x), dim=1)
+
+# --------------------------- Hooks -----------------------------------------
+
+
+def register_ddp_with_rpc_for_sparse_and_dense_hook(ddp_model, process_group, gs_rref, world_size):
+    """
+        DDP with CUDA RPC for both sparse and dense parameters.
+        hook_id = 1
+
+    """
+    sparse_future_container = []
+    dense_future_container = []
+
+    def rpc_for_sparse_and_dense_hook(state, bucket):
+        grad = bucket.get_tensors()[0]
+        if grad.is_sparse:
+            dim = grad.sparse_dim()
+            cpu_tensor = grad.to_dense().cpu()
+            sparse_future_container.append(
+                _remote_method(
+                    GradientServer.add_to_sparse_gradient,
+                    gs_rref,
+                    rank=rank,
+                    gradient=cpu_tensor,
+                    dim=dim,
+                )
+            )
+        else:
+            cpu_tensor = grad.cpu()
+            dense_future_container.append(
+                _remote_method(
+                    GradientServer.add_to_dense_gradient,
+                    gs_rref,
+                    rank=rank,
+                    gradient=cpu_tensor,
+                )
+            )
+        fut = torch.futures.Future()
+        fut.set_result(bucket.get_tensors()[0])
+        return fut
+
+    ddp_model.register_comm_hook(None, rpc_for_sparse_and_dense_hook)
+
+    return sparse_future_container, dense_future_container
+
+
+def register_dpp_with_rpc_for_sparse_nccl_allreduce_dense_hook(ddp_model, process_group, gs_rref, world_size):
+    """
+        DDP with CPU RPC for sparse parameters + NCCL AllReduce for dense parameters
+        hook_id = 2
+
+    """
+    sparse_future_container = []
+
+    def rpc_for_sparse_nccl_allreduce_dense_hook(state, bucket):
+        grad = bucket.get_tensors()[0]
+        if grad.is_sparse:
+            dim = grad.sparse_dim()
+            cpu_tensor = grad.to_dense()
+            sparse_future_container.append(
+                _remote_method(
+                    GradientServer.add_to_sparse_gradient,
+                    gs_rref,
+                    rank=rank,
+                    gradient=cpu_tensor,
+                    dim=dim,
+                )
+            )
+            fut = torch.futures.Future()
+            fut.set_result(bucket.get_tensors())
+            return fut
+        else:
+            tensors = [t / world_size for t in bucket.get_tensors()]
+            return process_group.allreduce(tensors).get_future()
+
+    ddp_model.register_comm_hook(None, rpc_for_sparse_nccl_allreduce_dense_hook)
+
+    return sparse_future_container
+
+
+def register_gloo_allreduce_for_sparse_and_nccl_allreduce_for_dense_hook(ddp_model, process_group, gs_rref, world_size):
```

Comment:
we can hold on a bit for the gloo version. Because, IIUC, the work returned by gloo allreduce does not yet support `get_future` yet. cc @agolynski @SciPioneer